# Chicago Crime Data Analysis

> This is by us. Handling crime data of Chicago from 2001 to 2017

## Import packages

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

## Reading Data

### Importing Data

In [2]:
Crime_2001_to_2004 = pd.read_csv("../Data/Crimes in Chicago_An extensive dataset of crimes in Chicago (2001-2017), by City of Chicago/Chicago_Crimes_2001_to_2004.csv", error_bad_lines=False)
Crime_2005_to_2007 = pd.read_csv("../Data/Crimes in Chicago_An extensive dataset of crimes in Chicago (2001-2017), by City of Chicago/Chicago_Crimes_2005_to_2007.csv", error_bad_lines=False)
Crime_2008_to_2011 = pd.read_csv("../Data/Crimes in Chicago_An extensive dataset of crimes in Chicago (2001-2017), by City of Chicago/Chicago_Crimes_2008_to_2011.csv", error_bad_lines=False)
Crime_2012_to_2017 = pd.read_csv("../Data/Crimes in Chicago_An extensive dataset of crimes in Chicago (2001-2017), by City of Chicago/Chicago_Crimes_2012_to_2017.csv", error_bad_lines=False)

b'Skipping line 1513591: expected 23 fields, saw 24\n'
C:\ProgramData\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3165: DtypeWarning: Columns (17,20) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,
b'Skipping line 533719: expected 23 fields, saw 24\n'
b'Skipping line 1149094: expected 23 fields, saw 41\n'


In [3]:
print(Crime_2001_to_2004.columns)
Crime_2001_to_2004.head()

Index(['Unnamed: 0', 'ID', 'Case Number', 'Date', 'Block', 'IUCR',
       'Primary Type', 'Description', 'Location Description', 'Arrest',
       'Domestic', 'Beat', 'District', 'Ward', 'Community Area', 'FBI Code',
       'X Coordinate', 'Y Coordinate', 'Year', 'Updated On', 'Latitude',
       'Longitude', 'Location'],
      dtype='object')


Unnamed: 0       ID Case Number                    Date  \
0         879  4786321    HM399414  01/01/2004 12:01:00 AM   
1        2544  4676906    HM278933  03/01/2003 12:00:00 AM   
2        2919  4789749    HM402220  06/20/2004 11:00:00 AM   
3        2927  4789765    HM402058  12/30/2004 08:00:00 PM   
4        3302  4677901    HM275615  05/01/2003 01:00:00 AM   

                 Block  IUCR                Primary Type  \
0    082XX S COLES AVE  0840                       THEFT   
1      004XX W 42ND PL  2825               OTHER OFFENSE   
2  025XX N KIMBALL AVE  1752  OFFENSE INVOLVING CHILDREN   
3   045XX W MONTANA ST  0840                       THEFT   
4   111XX S NORMAL AVE  0841                       THEFT   

                      Description Location Description  Arrest  ...  Ward  \
0   FINANCIAL ID THEFT: OVER $300            RESIDENCE   False  ...   7.0   
1         HARASSMENT BY TELEPHONE            RESIDENCE   False  ...  11.0   
2   AGG CRIM SEX ABUSE FAM MEMBER            RESIDENCE   False  ...  35.0   
3   FINANCIAL ID THEFT: OVER $300                OTHER   False  ...  31.0   
4  FINANCIAL ID THEFT:$300 &UNDER            RESIDENCE   False  ...  34.0   

   Community Area  FBI Code  X Coordinate  Y Coordinate    Year  \
0            46.0        06           NaN           NaN  2004.0   
1            61.0        26     1173974.0     1876757.0  2003.0   
2            22.0        20           NaN           NaN  2004.0   
3            20.0        06           NaN           NaN  2004.0   
4            49.0        06     1174948.0     1831051.0  2003.0   

               Updated On   Latitude  Longitude                       Location  
0  08/17/2015 03:03:40 PM        NaN        NaN                            NaN  
1  04/15/2016 08:55:02 AM  41.817229 -87.637328  (41.817229156, -87.637328162)  
2  08/17/2015 03:03:40 PM        NaN        NaN                            NaN  
3  08/17/2015 03:03:40 PM        NaN        NaN                            NaN  
4  04/15/2016 08:55:02 AM  41.691785 -87.635116  (41.691784636, -87.635115968)  

[5 rows x 23 columns]

### ERROR

When trying to read data with

`Crime_2001_to_2004 = pd.read_csv("../Data/Crimes in Chicago_An extensive dataset of crimes in Chicago (2001-2017), by City of Chicago/Chicago_Crimes_2001_to_2004.csv")`

Got an Error that goes like:

> ParserError: Error tokenizing data. C error: Expected 23 fields in line 1513591, saw 24

The solution is:

`Crime_2001_to_2004 = pd.read_csv("../Data/Crimes in Chicago_An extensive dataset of crimes in Chicago (2001-2017), by City of Chicago/Chicago_Crimes_2001_to_2004.csv", error_bad_lines=False)`

So now it works:

> b'Skipping line 1513591: expected 23 fields, saw 24\n'
/home/hanpeng/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (17,20) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,

The reason being:

> 读取文件时遇到和列数不对应的行，此时会报错。若报错行可以忽略，则添加该参数
> When the column number of one line doesn't met the header row's

### Data Combination

> To Combine the four files' data into one

In [5]:
Crime_2001_to_2017 = pd.concat([Crime_2001_to_2004, Crime_2005_to_2007, Crime_2008_to_2011, Crime_2012_to_2017])

In [ ]:
Crime_2001_to_2017.head()